File: leetcode/23.合并 K 个升序链表.py
ListNode.__lt__ = lambda a, b: a.val < b.val  # 让堆可以比较节点大小

class Solution:
    def mergetwolist(self, a, b):
        cura, curb = a, b
        head = ListNode(0)
        cur = head
        while cura and curb:
            if cura.val < curb.val:
                cur.next = cura
                cura = cura.next
            else:
                cur.next = curb
                curb = curb.next
            cur = cur.next
        cur.next = cura if cura else curb
        return head.next
    def merge(self, lists, l, r):
        if l == r: return lists[l]
        if l > r: return None
        mid = (l + r) // 2
        return self.mergetwolist(self.merge(lists, l, mid), self.merge(lists, mid+1, r))

    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        return self.merge(lists, 0, len(lists) - 1)

File: leetcode/23.合并 K 个升序链表.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeKLists(lists []*ListNode) *ListNode {
    return merge(lists, 0, len(lists)-1)
}

func merge(lists []*ListNode, l, r int) *ListNode{
    if l == r {return lists[l]} 
    if l > r {return nil}
    mid := (l + r) / 2
    return mergetwolist(merge(lists, l, mid), merge(lists, mid + 1, r))
}

func mergetwolist(a, b *ListNode) *ListNode{
    if a == nil && b == nil{
        return nil
    }
    var head = &ListNode{}
    dummy := head
    for a != nil && b != nil{
        if a. Val < b.Val{
            head.Next = a
            a = a.Next
        }else{
            head.Next = b
            b = b.Next
        }
        head = head.Next
    }
    if a != nil{
        head.Next = a
    }else if b != nil{
        head.Next = b
    }
    return dummy.Next
}